Step forward

alfoa opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)

- **Merged Diego's commits of Interface PP**
- **Fixed Outstream Printing and regolded faulty ROMs' tests**
- **Fixed ARMA surrogate model**
- **TEMPORARY fix of printing of HISTORY SET (Must be modified)**
- **Fixed Databases (loading, pushing, metadata)**
- **Fixed multi-target roms**
- **Added the possibility to ask for unfolded XArrays in the _realization(…)_**


##### What are the significant changes in functionality due to this change request?


----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

PaulTalbot-INL commented on framework/DataObjects/XHistorySet.py:
```diff
@@ -228,6 +238,15 @@ def _toCSV(self,fname,start=0,**kwargs):
+      ## always add the pivotParameter value in the printing.
+      ## Please check
+      ## rlz = data.isel(**{self.sampleTag:i})[ordered] (AA: actually this does not change anything)
+      rlz = data.isel(**{self.sampleTag:i})[ordered].dropna(self.indexes[0]) #
```
I don't think the intent was to drop the pivotParameter; on the contrary, the intent is to drop any pivot values that have NaN values, right?  Is this comment still relevant?

alfoa replied:
No, I got it later. The comment is not valid anymore and I am removing it

PaulTalbot-INL commented on framework/PostProcessorFunctions/HistorySetSnapShot.py:
```diff
@@ -190,106 +185,91 @@ def run(self,inputDic, pivotVal=None):
+  #  outputDic['data'][pivotParameter][rlz] = inputDic['data'][pivotParameter][rlz][timeStepID]
   return outputDic
+
+def returnIndexFirstPassage(array,value):
```
Doesn't numpy already have this method?

alfoa replied:
@mandd

PaulTalbot-INL commented on framework/PostProcessorFunctions/HistorySetSync.py:
```diff
@@ -1,20 +1,20 @@
 # Copyright 2017 Battelle Energy Alliance, LLC
 #
-# Licensed under the Apache License, Version 2.0 (the "License");
+# Licensed under the Apache License, Version 2.0 (the 'License');
```
Looks like an over-aggressive search-replace to me.

PaulTalbot-INL commented on framework/PostProcessorFunctions/TypicalHistoryFromHistorySet.py:
```diff
@@ -71,6 +77,190 @@ def run(self,inputDic):
+
+    ## Check if data is synchronized
+    referenceHistory = 0 # inputDict['output'].keys()[0]
+    referenceTimeAxis = inputDict[self.pivotParameter][referenceHistory]
```
We have a method in the dataobject for checking axial alignment.  Can we replace this with that?

mandd replied:
this was performed before this restructuring. I think we should maximize timing later.

PaulTalbot-INL commented on framework/PostProcessorFunctions/TypicalHistoryFromHistorySet.py:
```diff
@@ -71,6 +77,190 @@ def run(self,inputDic):
+
+    return outputDict
+
+  def run_OLD(self,inputDic):
```
Looks like leftover code.

mandd replied:
oh, yes. removed

PaulTalbot-INL commented on framework/PostProcessorFunctions/riskMeasuresDiscrete.py:
```diff
@@ -160,24 +159,32 @@ def run(self,inputDic):
 
+    outputDic['data']['ProbabilityWeight'] = np.asanyarray(1.0)
+    outputDic['data']['prefix'] = np.asanyarray(1.0)
+    print(outputDic)
```
debug print

mandd replied:
cleared

PaulTalbot-INL commented on framework/PostProcessorFunctions/riskMeasuresDiscrete.py:
```diff
@@ -160,24 +159,32 @@ def run(self,inputDic):
-    return outputDic
 
+    outputDic['data']['ProbabilityWeight'] = np.asanyarray(1.0)
+    outputDic['data']['prefix'] = np.asanyarray(1.0)
```
wish I would have known about this "asanyarray" sooner.  Seems quite useful.

PaulTalbot-INL commented on framework/PostProcessorInterfaceBaseClass.py:
```diff
@@ -109,6 +109,59 @@ def checkGeneratedDicts(self,outputDic):
+    return outcome
+
+
+  def checkOutputFormat_OLD(self,outputDic):
```
old code

mandd replied:
done

PaulTalbot-INL commented on framework/PostProcessors/ETImporter.py:
```diff
@@ -460,6 +460,7 @@ def collectOutput(self, finishedJob, output):
                                       'variables specified in the ET (' + str(set(outputDict.keys())))
     if output.type in ['PointSet']:
       for key in outputDict.keys():
+        print(type(output))
```
debug print

mandd replied:
removed

PaulTalbot-INL commented on framework/PostProcessors/InterfacedPostProcessor.py:
```diff
@@ -136,17 +136,47 @@ def run(self, inputIn):
+    #if self.postProcessor.checkGeneratedDicts(outputDic):
+    #  return outputDic
+    #else:
+    #  self.raiseAnError(RuntimeError,'InterfacedPostProcessor Post-Processor '+ self.name +' : function has generated a not valid output dictionary')
```
commented code

mandd replied:
removed

PaulTalbot-INL commented on framework/PostProcessors/InterfacedPostProcessor.py:
```diff
@@ -157,69 +187,66 @@ def collectOutput(self, finishedJob, output):
       self.raiseAnError(RuntimeError, "No available output to collect (run possibly not finished yet)")
-
     evaluation = evaluations[1]
+    output.addExpectedMeta(set(self.metaKeys))
```
Doesn't this conflict potentially with the hard-coded metavars kept in the individual IPPs?

mandd replied:
the IPP should not generate metavars

PaulTalbot-INL commented on framework/PostProcessors/InterfacedPostProcessor.py:
```diff
@@ -157,69 +187,66 @@ def collectOutput(self, finishedJob, output):
+    elif location=='output':
+      return self.outputFormat
+
+  def inputToInternal_OLD(self,inputs):
```
old code ... name?  Is this used or no?

mandd replied:
removed

PaulTalbot-INL commented on framework/PostProcessors/InterfacedPostProcessor.py:
```diff
@@ -157,69 +187,66 @@ def collectOutput(self, finishedJob, output):
     return inputDict
 
-  def inputToInternal(self,input):
+  def collectOutput(self, finishedJob, output):
```
Am I reading this right, that there's two `collectOutput` and two `returnFormat` defined in this class?

mandd replied:
I removed the useless ones

PaulTalbot-INL commented on framework/Samplers/Sampler.py:
```diff
@@ -580,6 +580,9 @@ def initialize(self,externalSeeding=None,solutionExport=None):
     for var in self.toBeSampled.keys():
+      #if self.variables2distributionsMapping[var]['totDim'] > 1:
+      #  meta +=  ['ProbabilityWeight-'+ var.replace(",","-")]
+      #else:
```
dead code

alfoa replied:
I left it there bacause what we were discussing today about the ND sampling ...that I honestly do not understand how we handle in the statistics (there is this replace for correlated dimensions but I do not see any special case in the basic statistics). @mandd and @wangcj05  any light here?

PaulTalbot-INL commented on framework/h5py_interface_creator.py:
```diff
@@ -275,7 +297,8 @@ def __populateGroup(self, group, name,  rlz):
       # get data names
-      group.create_dataset(name + "_dataOther", data=(json.dumps(np.concatenate( dataOther.values()).ravel().tolist())))
+      group.attrs[name + b'_dataOther'] = json.dumps(np.concatenate( dataOther.values()).ravel().tolist())
+      #group.create_dataset(name + "_dataOther", data=(json.dumps(np.concatenate( dataOther.values()).ravel().tolist())))
```
dead code

alfoa replied:
removed

PaulTalbot-INL commented on tests/framework/PostProcessors/InterfacedPostProcessor/test_interfacedPP.xml:
```diff
@@ -23,18 +23,17 @@
-      <Output class="DataObjects" type="PointSet">outMC</Output>
-      <Output class="Databases" type="HDF5">out_db</Output>
-      <Output class="OutStreams" type="Print">PrintHistorySet_dump</Output>
+      <Input   class="DataObjects" type="PointSet"      >inputPlaceHolder</Input>
```
This formatting will all get erased as soon as a conversion script is run.  I highly recommend against adding it.

I realize it's nice to see lined up, but I think we should leave that to a proper UI, not regression test files.

alfoa commented:
```
Number of success first file:
342
Number of success second file:
362
NUMBER OF SUCCESSES INCREASED: +20

```

wangcj05 commented on framework/PostProcessors/DataMining.py:
```diff
@@ -579,6 +578,74 @@ def collectOutput(self, finishedJob, outputObject):
+    for key in dataMineDict['outputs']:
+      if outputObject.type == 'PointSet':
+        outputObject.addVariable(key,dataMineDict['outputs'][key])
+      elif outputObject.type == 'HistorySet':
```
The following calculations seems to me are using the functions from the old DataObjects, such as the method 
```updateOutputValue```

mandd replied:
I dont think DataMining.py has been updated yet

wangcj05 replied:
Good.

wangcj05 commented on framework/PostProcessors/ETImporter.py:
```diff
@@ -460,6 +460,7 @@ def collectOutput(self, finishedJob, output):
                                       'variables specified in the ET (' + str(set(outputDict.keys())))
     if output.type in ['PointSet']:
       for key in outputDict.keys():
+        print(type(output))
```
remove print

mandd replied:
removed

wangcj05 commented on framework/PostProcessors/InterfacedPostProcessor.py:
```diff
@@ -136,17 +136,47 @@ def run(self, inputIn):
+    #if self.postProcessor.checkGeneratedDicts(outputDic):
+    #  return outputDic
+    #else:
+    #  self.raiseAnError(RuntimeError,'InterfacedPostProcessor Post-Processor '+ self.name +' : function has generated a not valid output dictionary')
```
remove commented lines

wangcj05 commented on framework/PostProcessors/InterfacedPostProcessor.py:
```diff
@@ -157,69 +187,66 @@ def collectOutput(self, finishedJob, output):
-
-
-  def inputToInternal_OLD(self,input):
+  def returnFormat(self,location):
```
add the docstring

mandd replied:
added

wangcj05 commented on framework/PostProcessors/InterfacedPostProcessor.py:
```diff
@@ -157,69 +187,66 @@ def collectOutput(self, finishedJob, output):
+    elif location=='output':
+      return self.outputFormat
+
+  def inputToInternal_OLD(self,inputs):
```
Should be removed?

wangcj05 commented on framework/Samplers/Sampler.py:
```diff
@@ -580,6 +580,9 @@ def initialize(self,externalSeeding=None,solutionExport=None):
     for var in self.toBeSampled.keys():
+      #if self.variables2distributionsMapping[var]['totDim'] > 1:
+      #  meta +=  ['ProbabilityWeight-'+ var.replace(",","-")]
+      #else:
```
Why this is commented out?

wangcj05 commented on tests/framework/PostProcessors/InterfacedPostProcessor/test_HS2PS.xml:
```diff
@@ -92,30 +87,6 @@
-          <text>Plot History</text>
-        </title>
-      </actions>
-    </Plot>
```
why plot is removed?

mandd replied:
I removed since it was not part of the testing capability

wangcj05 commented on tests/framework/PostProcessors/InterfacedPostProcessor/test_historySetSamplingIPP.xml:
```diff
@@ -157,10 +156,6 @@
-  <Databases>
-    <HDF5 name="out_db" readMode="overwrite"/>
-  </Databases>
-
```
why database is removed?

mandd replied:
I removed since it was not part of the testing capability

wangcj05 commented on tests/framework/PostProcessors/InterfacedPostProcessor/test_historySetSnapshot.xml:
```diff
@@ -149,10 +148,6 @@
-  <Databases>
-    <HDF5 name="out_db" readMode="overwrite"/>
-  </Databases>
-
```
why database is removed?

mandd replied:
I removed since it was not part of the testing capability

wangcj05 commented on tests/framework/PostProcessors/InterfacedPostProcessor/test_historySetSync.xml:
```diff
@@ -106,10 +105,6 @@
-  <Databases>
-    <HDF5 name="out_db" readMode="overwrite"/>
-  </Databases>
-
```
same here?

mandd replied:
I removed since it was not part of the testing capability

wangcj05 commented on tests/framework/PostProcessors/InterfacedPostProcessor/test_interfacedPP.xml:
```diff
@@ -80,10 +79,6 @@
 
-  <Databases>
-    <HDF5 name="out_db" readMode="overwrite"/>
-  </Databases>
```
Could you provide a reason why the database is removed?

mandd replied:
I removed since it was not part of the testing capability

wangcj05 commented on tests/framework/PostProcessors/InterfacedPostProcessor/test_interfacedPP_pointset.xml:
```diff
@@ -79,10 +78,6 @@
-  <Databases>
-    <HDF5 name="out_db" readMode="overwrite"/>
-  </Databases>
-
```
same as above.

mandd replied:
I removed since it was not part of the testing capability

alfoa commented on framework/Models/Dummy.py:
```diff
@@ -105,7 +104,8 @@ def _inputToInternal(self,dataIN,full=False):
               if localInput[entries] is None:
                 localInput[entries] = []
-              localInput[entries].append(np.full((sizeIndex,),dataSet.isel(RAVEN_sample_ID=hist)[entries].values))
+              value = dataSet.isel(RAVEN_sample_ID=hist)[entries].values
```
addressed

alfoa commented on framework/PostProcessorFunctions/HistorySetSnapShot.py:
```diff
@@ -136,51 +134,48 @@ def readMoreXML(self,xmlNode):
-
       #for timeSlice we call historySetWindow
       if self.type == 'timeSlice':
         outputHSDic = self.HSsyncPP.run([inputDic])
```
@mandd  This is a question for you

alfoa commented on framework/PostProcessorFunctions/HistorySetSnapShot.py:
```diff
@@ -190,106 +185,91 @@ def run(self,inputDic, pivotVal=None):
-      for keys in outVars:
+  outputDic={'data':{}}
+  outputDic['data']['ProbabilityWeight'] = inputDic['data']['ProbabilityWeight']
+  outputDic['data']['prefix'] = inputDic['data']['prefix']
```
@mandd

PaulTalbot-INL replied:
OH I misread; now I understand you're not trying to preserve metadata here, you're actually setting the probability weight and prefix.  I appreciate that more now.

alfoa commented on framework/DataObjects/XDataSet.py:
```diff
@@ -1513,10 +1522,25 @@ def _usePandasWriteCSV(self,fName,data,ordered,keepSampleTag=False,keepIndex=Fal
-        if indexx:
-          data.index = data.index.droplevel(self.sampleTag)
-        data.to_csv(fName+'.csv',mode=mode,header=header, index=indexx)
+        indexx = True
```
done

alfoa commented on framework/DataObjects/XDataSet.py:
```diff
@@ -1513,10 +1522,25 @@ def _usePandasWriteCSV(self,fName,data,ordered,keepSampleTag=False,keepIndex=Fal
+        if not indexx:
+          data.to_csv(fName+'.csv',mode=mode,header=header, index=indexx)
+        else:
+          ## This is extremely bad and not elegant
```
added

alfoa commented on framework/DataObjects/XHistorySet.py:
```diff
@@ -153,9 +153,16 @@ def _fromCSV(self,fname,**kwargs):
+      subFile = sub
+      # check if the sub has an absolute path, otherwise take it from the master file (fname)
+      if not os.path.isabs(subFile):
+        subFile = os.path.join(os.path.dirname(fname),subFile)
```
replaced everywhere

PaulTalbot-INL replied:
For the record, I don't necessarily agree with this idea.  "fname" is an abbreviation of "filename" which is one word.  Anyway, fine that it's changed, but more unnecessary camelback tends to make the code less readable in my opinion.

alfoa commented on framework/DataObjects/XHistorySet.py:
```diff
@@ -173,6 +180,9 @@ def _identifyVariablesInCSV(self,fname):
       subfile = base.readline().split(',')[-1].strip()
+      # check if abs path otherwise take the dirpath from the master file (fname)
+      if not os.path.isabs(subfile):
+        subfile = os.path.join(os.path.dirname(fname),subfile)
```
fixed

alfoa commented on framework/PostProcessorFunctions/HS2PS.py:
```diff
@@ -83,65 +82,48 @@ def run(self,inputDic):
+      historyLength = len(inputDict['data'][self.features[0]])
+      numVariables = historyLength*len(self.features)
+      for history in inputDict['data'][self.features[0]]:
+        if len(history) != historyLength:
```
@mandd

alfoa commented:
@PaulTalbot-INL @wangcj05 I addressed the comments that are not part of the InterfacedPostProcessors. I think the Interface PPs are not final yet and @mandd will work for finalizing them. Now we have 2 options:
- we merge this in dataobject-rework and we re-branch from there for our own parallel development
- we branch from this branch and use this as additional buffer.

Thoughts?

PaulTalbot-INL commented on framework/PostProcessorFunctions/HistorySetSnapShot.py:
```diff
@@ -190,106 +185,91 @@ def run(self,inputDic, pivotVal=None):
+  index=-1
+  for i in range(1,array.size):
+    if (array[i]>=value and array[i-1]<=value) or (array[i]<=value and array[i-1]>=value):
+      index = i
```
I agree.

I will note, however, almost everything we do in interfaced postprocessors can be done much more efficiently via direct interaction with the DataObject, skipping the dictionary form.  If we're going to improve efficiency, let's go big with it (and save it for after we get tests green).

PaulTalbot-INL commented on framework/PostProcessorFunctions/TypicalHistoryFromHistorySet.py:
```diff
@@ -71,6 +77,190 @@ def run(self,inputDic):
+    #get actual data
+    inputDict = inputDic[0]['data']
+    #identify features
+    self.features = inputDict['outVars']# self.features = inputDict['output'][inputDict['output'].keys()[0]].keys()
```
Can we save a screenshot of these two lines just to show how much simpler the new interface is? 😄  Just kidding.  Still, though, I like how short the new API is.

PaulTalbot-INL commented:
I vote for merging this branch in and branching off of that.  We need to minimize the number of "on my branch" improvements to prevent re-fixing as much as possible, in my opinion.

wangcj05 commented:
Personally, I prefer the first option. But I think we should address all the comments relating to the interfacePP before this can be merged.

alfoa commented:
the ```dataobejct-rework``` is our "working" branch. I think we should merge this and everybody can branch from the ```dataobejct-rework``` otherwise (since the InterfacePP comments are quite heavy to be addressed) we are going to wait to long to continue our development

PaulTalbot-INL commented:
How are we going to preserve all the comments here?  I'm not sure what the final merge process is going to look like, but I don't think it will be straightforward to review the entire code change with as much rigor as we've applied to this particular PR.

mandd commented on framework/PostProcessorFunctions/TypicalHistoryFromHistorySet.py:
```diff
@@ -71,6 +77,190 @@ def run(self,inputDic):
+    # task: reshape the data into histories with the size of the output I'm looking for
+    #data dictionaries have form {historyNumber:{VarName:[data], VarName:[data]}}
+    reshapedData = {}
+    #keyNewH = 0
```
done

mandd commented on framework/PostProcessorFunctions/TypicalHistoryFromHistorySet.py:
```diff
@@ -71,6 +77,190 @@ def run(self,inputDic):
+    reshapedData = {}
+    #keyNewH = 0
+    newHistoryCounter = 0 #new history tracking labels
+    for historyNumber in range(inputDic['numberRealizations']):  #inputDict['output'].keys():
```
done

mandd commented on framework/PostProcessorFunctions/TypicalHistoryFromHistorySet.py:
```diff
@@ -71,6 +77,190 @@ def run(self,inputDic):
+    for var in typicalData.keys():
+      outputDict['data'][var] = typicalData[var]
+    # preserve input data
+    #outputDict['data']['input'][1] = dict((keyIn,np.array(inputDict['input'].values()[0][keyIn][0])) for keyIn in inputDict['input'].values()[0].keys())
```
cleared

mandd commented on framework/PostProcessorFunctions/riskMeasuresDiscrete.py:
```diff
@@ -160,24 +159,32 @@ def run(self,inputDic):
 
+    outputDic['data']['ProbabilityWeight'] = np.asanyarray(1.0)
+    outputDic['data']['prefix'] = np.asanyarray(1.0)
+    print(outputDic)
```
I added this part here as a place-holder for when we will extend the capabilities for this PP to combine analysis from multiple data sets

alfoa commented:
This can be merged and everybody should fork/branch from the dataobject-rework branch

mandd commented:
I need a bit more time to address the comments

mandd commented on framework/PostProcessorFunctions/riskMeasuresDiscrete.py:
```diff
@@ -190,6 +197,159 @@ def runStatic(self,inputDic, componentConfig=None):
+          pbWeights  = np.ones(pointCount)/float(pointCount)
+
+        if inputName in self.IEdata.keys():
+          multiplier = self.IEdata[inputName]
```
edited

mandd commented on framework/PostProcessorFunctions/riskMeasuresDiscrete.py:
```diff
@@ -190,6 +197,159 @@ def runStatic(self,inputDic, componentConfig=None):
+            ## R0 = pb of system failure
+            R0     = np.sum(pbWeights[indexSystemFailure])
+            Rminus = np.sum(pbWeights[indexFailureMinus]) / np.sum(pbWeights[indexComponentMinus])
+            Rplus  = np.sum(pbWeights[indexFailurePlus]) / np.sum(pbWeights[indexComponentPlus])
```
done

wangcj05 commented:
@mandd Can we merge this? or do you want us to wait?

mandd commented on framework/PostProcessorFunctions/riskMeasuresDiscrete.py:
```diff
@@ -190,6 +197,159 @@ def runStatic(self,inputDic, componentConfig=None):
+
+    ## If for whatever reason passing an empty input back causes errors, then you may want to add some sort of dummy
+    ## value.
+    # outputDic['data']['input'] = {} # {'dummy' : np.asanyarray(0)}
```
edited

mandd commented on framework/PostProcessorFunctions/riskMeasuresDiscrete.py:
```diff
@@ -190,6 +197,159 @@ def runStatic(self,inputDic, componentConfig=None):
+
+    return outputDic
+
+  def runStatic_OLD(self,inputDic, componentConfig=None):
```
done

mandd commented on framework/PostProcessorFunctions/testInterfacedPP_PointSet.py:
```diff
@@ -19,6 +19,8 @@
 warnings.simplefilter('default',DeprecationWarning)
 
+import copy
+
```
done

mandd commented on framework/PostProcessorInterfaceBaseClass.py:
```diff
@@ -109,6 +109,59 @@ def checkGeneratedDicts(self,outputDic):
+    return outcome
+
+
+  def checkOutputFormat_OLD(self,outputDic):
```
yes

wangcj05 commented on framework/PostProcessorFunctions/HS2PS.py:
```diff
@@ -95,9 +95,9 @@ def run(self,inputDic):
+      historyLength = len(inputDict['data'][self.features[0]][0])
       numVariables = historyLength*len(self.features)
-      for history in inputDict['data'][self.features[0]]:
+      for history in inputDict['data'][self.features[0]][history]:
```
I think the ```history``` at the end should be removed.

PaulTalbot-INL commented:
My comments have been sufficiently addressed.

wangcj05 commented:
This PR will be merged. Everything looks fine to me.
